servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfig.java

```diff
@@ -0,0 +1,124 @@
+ * limitations under the License.
+ */
+package io.servicetalk.http.api;
+
+import java.util.List;
+
+/**
+ * Configuration options for <a href="https://datatracker.ietf.org/doc/html/rfc7231#section-6.4">redirection</a>.
+ */
+public interface RedirectConfig {
+
+    /**
+     * Predicate to make the final decision if redirect should be performed or not based on the given context.
+     * <p>
+     * Implementations should prefer running this predicate as the last check, after validation of the
+     * {@link HttpResponseStatus}, {@link RedirectConfig#maxRedirects()}, {@link RedirectConfig#allowedMethods()},
+     * presence of the {@link HttpHeaderNames#LOCATION Location} header, and
+     * {@link RedirectConfig#allowNonRelativeRedirects()}.
+     */
+    @FunctionalInterface
+    interface ShouldRedirectPredicate {
+
+        /**
+         * Decides if a redirect should be performed or not based on the given context.
+         *
+         * @param relative if {@code true}, the redirect location was identified as relative to the previous request
+         * @param redirectCount sequential counter of already processed redirects (starts from {@code 0})
+         * @param previousRequest previous request that was redirected
+         * @param redirectResponse response to redirect
+         * @return {@code true} if the redirect should be processed based on the given context
+         */
+        boolean test(boolean relative, int redirectCount,
+                     HttpRequestMetaData previousRequest, HttpResponseMetaData redirectResponse);
+    }
+
+    /**
+     * Provides access to the full context of the redirect to apply transformations for the pre-initialized redirect
+     * request.
+     * <p>
+     * It can be used to add/remove headers, payload body, or trailers.
+     */
+    @FunctionalInterface
+    interface RedirectRequestTransformer {
+
+        /**
+         * Applies transformations for the pre-initialized redirect request.
+         *
+         * @param relative if {@code true}, the redirect location was identified as relative to the previous request
+         * @param previousRequest previous request that was redirected
+         * @param redirectResponse response to redirect
+         * @param redirectRequest pre-initialized request to follow the redirect
+         * @return the final {@link StreamingHttpRequest} that will be send to follow the redirect after all
+         * transformations applied
+         */
+        StreamingHttpRequest apply(boolean relative, StreamingHttpRequest previousRequest,
+                                   StreamingHttpResponse redirectResponse, StreamingHttpRequest redirectRequest);
+    }
+
+    /**
+     * Maximum number of redirects to follow.
+     *
+     * @return maximum number of redirects to follow.
+     */
+    int maxRedirects();
+
+    /**
+     * {@link HttpRequestMethod}s that are allowed to follow redirects.
```

Comment:
the current implementation requires the list is sorted, and doesn't verify this constraint. consider the following options:
- clarify this constraint in the javadoc, in `RedirectSingle` check if they are not sorted and throw
- in `RedirectSingle` just sort them if they are not sorted.
- make this a responsibility of `RedirectRequestTransformer` which can internalize sorting requirement

Reply:
Good point, I've changed it to `Set` to avoid any ambiguity and clarify on the API that duplicates are not expected. Not sure if a `SortedSet` will provide any value.